packages/@lwc/engine-dom/src/custom-elements/create-scoped-registry.ts

```diff
@@ -0,0 +1,524 @@
+        const instance = upgradingInstance;
+        if (!isUndefined(instance)) {
+            upgradingInstance = undefined;
+            return instance;
+        }
+        // Construction case: we need to construct the pivoting instance and return it.
+        // This is possible when the user register it via global registry and instantiate
+        // it via `new Ctor()`.
+        const { constructor } = this;
+        const definition = globalDefinitionsByClass.get(constructor as CustomElementConstructor);
+        if (isUndefined(definition) || isUndefined(definition.PivotCtor)) {
+            // This code path is hit if someone `new`s a class that extends `HTMLElement` without
+            // doing `customElements.define()` first. This matches native browser behavior:
+            // https://stackoverflow.com/a/61883392
+            throw new TypeError('Illegal constructor');
+        }
+        // This constructor is ONLY invoked when it is the user instantiating
+        // an element via new Ctor while Ctor is a registered global constructor.
+        const { PivotCtor, UserCtor } = definition;
+        return new PivotCtor(UserCtor);
+    };
+    HTMLElement.prototype = NativeHTMLElement.prototype;
+
+    /**
+     * Create a new PivotConstructor for the given tagName, which is capable of being constructed
+     * with a UserConstructor defining the behavior. Passing in the UserConstructor here
+     * is a hint that can be used when registering a custom element with the global custom elements
+     * registry for the first time, which provides certain optimizations. It also marks the UserConstructor
+     * as "safe" to be used when passed in to a PivotConstructor.
+     *
+     * @param tagName - element tag name
+     * @param UserCtor - userland custom element constructor
+     * @returns a new custom element constructor
+     */
+    return function createPivotConstructor(
+        tagName: string,
+        UserCtor: CustomElementConstructor
+    ): CustomElementConstructor {
+        tagName = StringToLowerCase.call(tagName);
+        let PivotCtor = pivotCtorByTag.get(tagName);
+        if (isUndefined(PivotCtor)) {
+            const definition = getOrCreateDefinitionForConstructor(UserCtor);
+            PivotCtor = createPivotingClass(tagName, definition);
+            // Register a pivoting class as a global custom element
+            nativeDefine.call(nativeRegistry, tagName, PivotCtor);
+            definition.PivotCtor = PivotCtor;
+            // Only cache after nativeDefine has been called, because if it throws an error
+            // (e.g. for an invalid tag name), then we don't want to cache anything.
+            definitionForConstructor.set(UserCtor, definition);
+            pivotCtorByTag.set(tagName, PivotCtor);
+        }
+        // Register a UserConstructor as "safe" to be used within a PivotConstructor
+        registeredUserCtors.add(UserCtor);
+        return PivotCtor;
+    };
+}
```

Comment:
GitHub fails at giving a good diff here because I renamed the file. But basically I wrapped the old `createPivotConstructor` in a function that returns the `createPivotConstructor` function. Eventually I want to move this to a separate package so Locker can reuse it.